mkosi/backend.py

```diff
@@ -270,23 +280,86 @@ def read(cls, loopdev: Path) -> PartitionTable:
     def empty(cls, first_lba: Optional[int] = None) -> PartitionTable:
         return cls([], None, 512, first_lba)
 
-    def first_usable_offset(self, max_partitions: int = 128) -> int:
-        if self.last_partition_sector:
-            return roundup(self.last_partition_sector, self.grain)
-        elif self.first_lba is not None:
+    def first_partition_offset(self, max_partitions: int = 128) -> int:
+        if self.first_lba is not None:
             # No rounding here, we honour the specified value exactly.
             return self.first_lba * self.sector_size
         else:
             # The header is like the footer, but we have a one-sector "protective MBR" at offset 0
             return roundup(self.sector_size + self.footer_size(), self.grain)
 
+    def last_partition_offset(self, max_partitions: int = 128) -> int:
+        if self.last_partition_sector:
+            return roundup(self.last_partition_sector * self.sector_size, self.grain)
+        else:
+            return self.first_partition_offset(max_partitions)
+
     def footer_size(self, max_partitions: int = 128) -> int:
         # The footer must have enough space for the GPT header (one sector),
         # and the GPT parition entry area. PEA size of 16384 (128 partitions)
         # is recommended.
         pea_sectors = math.ceil(max_partitions * 128 / self.sector_size)
         return (1 + pea_sectors) * self.sector_size
 
+    def disk_size(self) -> int:
+        return roundup(self.last_partition_offset() + self.footer_size(), self.grain)
+
+    def add(self,
+            label: str,
+            size: int,
+            type_uuid: uuid.UUID,
+            name: str,
+            part_uuid: Optional[uuid.UUID] = None,
+            read_only: Optional[bool] = False) -> Partition:
+
+        assert '"' not in name
+
+        size = roundup(size, self.grain)
+        n_sectors = size // self.sector_size
+        new = [f'size={size // self.sector_size}',
+               f'type={type_uuid}',
+               f'attrs={"GUID:60" if read_only else ""}',
+               f'name="{name}"',
+               f'uuid={part_uuid}' if part_uuid is not None else None]
+
+        self.partitions += [', '.join(filter(None, new))]
+        self.last_partition_sector = self.last_partition_offset() // self.sector_size + n_sectors
+
+        part = Partition(size, len(self.partitions))
+
+        if label is not None:
```

Comment:
why make this optional? shouldn't this be a requirement?

if this was required then we could just store the entries in one data structure instead of two